libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+  ## - `IdentityPubKeySerializationError` if serialization of identity public key fails.
+  ## - `IdentitySigningError` if signing the hash fails.
+  ## - `ASN1EncodingError` if ASN.1 encoding fails.
+
+  # Serialize the Certificate's Public Key
+  var
+    certPubKeyDer: array[512, byte]
+    certPubKeyDerLen: cint
+
+  certPubKeyDerLen = mbedtls_pk_write_pubkey_der(
+    unsafeAddr certificateKeypair, addr certPubKeyDer[0], certPubKeyDer.len.uint
+  )
+  if certPubKeyDerLen < 0:
+    raise newException(
+      CertificateCreationError, "Failed to write certificate public key in DER format"
+    )
+
+  # Adjust pointer to the start of the data
+  let certPubKeyDerPtr = addr certPubKeyDer[certPubKeyDer.len - certPubKeyDerLen]
+
+  # Create the Message to Sign
+  var msg = newSeq[byte](P2P_SIGNING_PREFIX.len + certPubKeyDerLen.int.int)
+
+  # Copy the prefix into msg
+  for i in 0 ..< P2P_SIGNING_PREFIX.len:
+    msg[i] = byte(P2P_SIGNING_PREFIX[i])
+
+  # Copy the public key DER into msg
+  copyMem(addr msg[P2P_SIGNING_PREFIX.len], certPubKeyDerPtr, certPubKeyDerLen.int)
+
+  # Compute SHA-256 hash of the message
+  var hash: array[32, byte]
+  let hashRet = mbedtls_sha256(
+    msg[0].addr, msg.len.uint, addr hash[0], 0 # 0 for SHA-256
+  )
+  if hashRet != 0:
+    # Since hashing failure is critical and unlikely, we can raise a general exception
+    raise newException(TLSCertificateError, "Failed to compute SHA-256 hash")
+
+  # Sign the hash with the Identity Key
+  let signatureResult = identityKeypair.seckey.sign(hash)
+  if signatureResult.isErr:
+    raise newException(
+      IdentitySigningError, "Failed to sign the hash with the identity key"
+    )
+  let signature = signatureResult.get().data
+
+  # Get the public key bytes
+  let pubKeyBytesResult = identityKeypair.pubkey.getBytes()
+  if pubKeyBytesResult.isErr:
+    raise newException(
+      IdentityPubKeySerializationError, "Failed to get identity public key bytes"
+    )
+  let pubKeyBytes = pubKeyBytesResult.get()
+
+  # Generate the SignedKey ASN.1 structure
+  return generateSignedKey(signature, pubKeyBytes)
+
+proc generate*(
+    identityKeyPair: KeyPair
+): (seq[byte], seq[byte]) {.
+    raises: [
+      KeyGenerationError, CertificateCreationError, ASN1EncodingError,
+      IdentityPubKeySerializationError, IdentitySigningError, TLSCertificateError,
+    ]
+.} =
+  ## Generates a self-signed X.509 certificate with the libp2p extension.
+  ##
+  ## Parameters:
+  ## - `identityKeyPair`: The peer's identity key pair.
+  ##
+  ## Returns:
+  ## A tuple containing:
+  ## - The DER-encoded certificate.
+  ## - The DER-encoded private key.
+  ##
+  ## Raises:
+  ## - `KeyGenerationError` if key generation fails.
+  ## - `CertificateCreationError` if certificate creation fails.
+  ## - `ASN1EncodingError` if encoding fails.
+  # Initialize entropy and DRBG contexts
+  var
+    entropy = mbedtls_entropy_context()
```

Comment:
If I remember correctly, entropy is a bit heavy to create. It might be interesting to find a way to only initialize it once. I'm not 100% sure though, but this could be checked.

Reply:
And how would be the right way of initializing and freeing this only once in Nim?

Reply:
I tried one approach.